File: mailblastr-ruby/lib/mailblastr.rb
# frozen_string_literal: true

require "mailblastr/version"
require "mailblastr/error"
require "mailblastr/client"
require "mailblastr/emails"
require "mailblastr/domains"
require "mailblastr/audiences"
require "mailblastr/contacts"
require "mailblastr/contact_properties"
require "mailblastr/segments"
require "mailblastr/topics"
require "mailblastr/campaigns"
require "mailblastr/templates"
require "mailblastr/automations"
require "mailblastr/webhooks"
require "mailblastr/events"
require "mailblastr/api_keys"
require "mailblastr/logs"
require "mailblastr/polls"

# The MailBlastr API client.
#
#   Mailblastr.api_key = "mb_xxxxxxxxx"
#
#   sent = Mailblastr::Emails.send({
#     from: "Acme <hello@yourdomain.com>",
#     to: ["user@example.com"],
#     subject: "Hello from MailBlastr",
#     html: "<p>Your first email</p>"
#   })
#   sent["id"] # => "..."
module Mailblastr
  DEFAULT_BASE_URL = "https://api.mailblastr.com"

  # Per-request network timeout, in seconds, applied to both the connect
  # (open) and read phases. 0 or nil means "no timeout".
  DEFAULT_TIMEOUT = 30

  # How many times a retryable response (HTTP 429/503) is retried before
  # giving up. 0 disables retries (a single attempt).
  DEFAULT_MAX_RETRIES = 2

  class << self
    # Your API key, e.g. "mb_xxxxxxxxx". Required before any call.
    attr_accessor :api_key

    # Override the API host (defaults to https://api.mailblastr.com).
    attr_writer :base_url

    # Per-request timeout in seconds (default 30). Set 0 (or nil) for none.
    attr_writer :timeout

    # Max automatic retries on HTTP 429/503 (default 2, i.e. up to 3 tries).
    attr_writer :max_retries

    def base_url
      @base_url || DEFAULT_BASE_URL
    end

    def timeout
      @timeout.nil? ? DEFAULT_TIMEOUT : @timeout
    end

    def max_retries
      @max_retries.nil? ? DEFAULT_MAX_RETRIES : @max_retries
    end

    #   Mailblastr.configure do |config|
    #     config.api_key = ENV["MAILBLASTR_API_KEY"]
    #   end
    def configure
      yield self
    end
  end
end

File: mailblastr-java/src/main/java/com/mailblastr/Mailblastr.java
package com.mailblastr;

import com.mailblastr.http.ApiClient;
import com.mailblastr.http.DefaultHttpTransport;
import com.mailblastr.http.HttpTransport;
import com.mailblastr.resources.ApiKeys;
import com.mailblastr.resources.Audiences;
import com.mailblastr.resources.Automations;
import com.mailblastr.resources.Batch;
import com.mailblastr.resources.Campaigns;
import com.mailblastr.resources.ContactProperties;
import com.mailblastr.resources.Contacts;
import com.mailblastr.resources.Domains;
import com.mailblastr.resources.Emails;
import com.mailblastr.resources.Events;
import com.mailblastr.resources.Logs;
import com.mailblastr.resources.Polls;
import com.mailblastr.resources.Segments;
import com.mailblastr.resources.Templates;
import com.mailblastr.resources.Topics;
import com.mailblastr.resources.Webhooks;

import java.time.Duration;

/**
 * The MailBlastr API client.
 *
 * <pre>{@code
 * Mailblastr mailblastr = new Mailblastr("mb_xxxxxxxxx");
 *
 * SendEmailRequest req = SendEmailRequest.builder()
 *     .from("Acme <hi@yourdomain.com>")
 *     .to("a@b.com")
 *     .subject("hello")
 *     .html("<p>hi</p>")
 *     .build();
 *
 * MailblastrResponse res = mailblastr.emails().send(req);
 * String id = res.getString("id");
 * }</pre>
 *
 * <p>Every method returns a {@link MailblastrResponse} (binary downloads
 * return {@code byte[]}); non-2xx responses throw {@link MailblastrException}.
 *
 * <p>Every request carries {@code Authorization: Bearer <your mb_ key>} and a
 * non-empty {@code User-Agent} — the API rejects a request without the latter
 * with HTTP 403 {@code validation_error}, before it even authenticates.
 */
public class Mailblastr {
    public static final String DEFAULT_BASE_URL = "https://www.mailblastr.com/api";
    public static final String VERSION = "5.0.1";
    public static final String USER_AGENT = "mailblastr-java/" + VERSION;

    /**
     * Longest {@code Idempotency-Key} the API accepts. The accepted range is
     * <strong>1&ndash;255</strong> characters measured after the server trims the
     * value (the storage column is {@code VARCHAR(255)}) — not 256. Anything
     * outside it is a {@code 400 invalid_idempotency_key}.
     *
     * <p>The header is honoured by {@code POST /emails} and
     * {@code POST /emails/batch} ONLY. Every other endpoint ignores it, so a
     * retry there creates a second resource.
     *
     * <p>This SDK does not check the length itself — the server is the
     * authority. The constant is exported so the rule is discoverable.
     */
    public static final int IDEMPOTENCY_KEY_MAX_LENGTH = 255;

    private final Emails emails;
    private final Batch batch;
    private final Domains domains;
    private final Audiences audiences;
    private final Contacts contacts;
    private final ContactProperties contactProperties;
    private final Campaigns campaigns;
    private final Segments segments;
    private final Topics topics;
    private final Templates templates;
    private final Automations automations;
    private final Webhooks webhooks;
    private final Logs logs;
    private final Events events;
    private final ApiKeys apiKeys;
    private final Polls polls;

    public Mailblastr(String apiKey) {
        this(apiKey, DEFAULT_BASE_URL);
    }

    /** Override the API host, e.g. for a self-hosted deployment. */
    public Mailblastr(String apiKey, String baseUrl) {
        this(apiKey, baseUrl, new DefaultHttpTransport());
    }

    /**
     * Override the API host and the HTTP-core robustness policy.
     *
     * @param timeout    per-request (and connect) timeout; {@code null} or a non-positive
     *                   value means "no timeout". Defaults to 30 seconds.
     * @param maxRetries extra automatic attempts on HTTP 429/503 (0 disables retries).
     *                   Defaults to 2 (up to 3 total attempts).
     */
    public Mailblastr(String apiKey, String baseUrl, Duration timeout, int maxRetries) {
        this(apiKey, baseUrl, new DefaultHttpTransport(timeout, maxRetries));
    }

    /** Inject a custom {@link HttpTransport} (used by the test suite). */
    public Mailblastr(String apiKey, String baseUrl, HttpTransport transport) {
        ApiClient api = new ApiClient(apiKey, baseUrl, USER_AGENT, transport);
        this.emails = new Emails(api);
        this.batch = new Batch(api);
        this.domains = new Domains(api);
        this.audiences = new Audiences(api);
        this.contacts = new Contacts(api);
        this.contactProperties = new ContactProperties(api);
        this.campaigns = new Campaigns(api);
        this.segments = new Segments(api);
        this.topics = new Topics(api);
        this.templates = new Templates(api);
        this.automations = new Automations(api);
        this.webhooks = new Webhooks(api);
        this.logs = new Logs(api);
        this.events = new Events(api);
        this.apiKeys = new ApiKeys(api);
        this.polls = new Polls(api);
    }

    /** Sent email (send / batch / list / get / attachments / reschedule / cancel) + {@code emails().receiving()}. */
    public Emails emails() { return emails; }

    /** Batch send — alias surface for {@code POST /emails/batch}. */
    public Batch batch() { return batch; }

    /** Sending/receiving domains, incl. claims and one-click DNS applies. */
    public Domains domains() { return domains; }

    public Audiences audiences() { return audiences; }

    /** Domain-first contacts (flat + nested APIs, batch/CSV import, segments, topics). */
    public Contacts contacts() { return contacts; }

    /** Custom contact properties (merge tags). */
    public ContactProperties contactProperties() { return contactProperties; }

    /** Domain-first campaigns (bulk sends, A/B, follow-ups, recurrence). */
    public Campaigns campaigns() { return campaigns; }

    /** Domain-first segments. */
    public Segments segments() { return segments; }

    /** Domain-first topics (granular subscription preferences). */
    public Topics topics() { return topics; }

    public Templates templates() { return templates; }

    /** Domain-first automations (steps, connections, runs). */
    public Automations automations() { return automations; }

    /** Webhooks, incl. the local {@code verifyWebhookSignature} helper. */
    public Webhooks webhooks() { return webhooks; }

    /** API request logs (read-only). */
    public Logs logs() { return logs; }

    /** Custom automation events — {@code domain} is REQUIRED on send. */
    public Events events() { return events; }

    public ApiKeys apiKeys() { return apiKeys; }

    /** Read-only in-email poll results. */
    public Polls polls() { return polls; }
}
